src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    helpers::Tasks,
+    Data,
+    Environment,
+    LedgerReader,
+    LedgerRequest,
+    LedgerRouter,
+    Message,
+    NodeType,
+    PeersRequest,
+    PeersRouter,
+    ProverRouter,
+};
+use snarkos_storage::{storage::Storage, OperatorState};
+use snarkvm::{algorithms::crh::sha256d_to_u64, dpc::prelude::*, utilities::ToBytes};
+
+use anyhow::Result;
+use rand::thread_rng;
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    path::Path,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tokio::{
+    sync::{mpsc, oneshot, RwLock},
+    task,
+    task::JoinHandle,
+};
+
+/// Shorthand for the parent half of the `Operator` message channel.
+pub(crate) type OperatorRouter<N> = mpsc::Sender<OperatorRequest<N>>;
+#[allow(unused)]
+/// Shorthand for the child half of the `Operator` message channel.
+type OperatorHandler<N> = mpsc::Receiver<OperatorRequest<N>>;
+
+///
+/// An enum of requests that the `Operator` struct processes.
+///
+#[derive(Debug)]
+pub enum OperatorRequest<N: Network> {
+    /// PoolRegister := (peer_ip, worker_address)
+    PoolRegister(SocketAddr, Address<N>),
+    /// PoolResponse := (peer_ip, proposed_block, worker_address)
+    PoolResponse(SocketAddr, Block<N>, Address<N>),
+}
+
+/// The predefined base share difficulty.
+const BASE_SHARE_DIFFICULTY: u64 = u64::MAX / 2u64.pow(1);
+
+///
+/// An operator for a program on a specific network in the node server.
+///
+#[derive(Debug)]
+pub struct Operator<N: Network, E: Environment> {
+    /// The address of the operator.
+    address: Option<Address<N>>,
```

Comment:
can the operator not have an Aleo address?

Reply:
It needs one, as the coinbase should be attributed to it.

Reply:
As we do instantiate the component regardless of node type, so in some cases there will be no address available.